File: backend/src/services/auditService.js
const prisma = require('../config/db');

/**
 * Creates an immutable audit log entry.
 * Fails safely if logging encounters a transient error to avoid crashing the main transaction
 * unless strictly required by compliance configuration.
 */
async function logAudit({ userId, action, resource, details, ipAddress, prismaClient = prisma }) {
  try {
    return await prismaClient.auditLog.create({
      data: {
        userId: userId || null,
        action,
        resource,
        details: typeof details === 'object' ? JSON.stringify(details) : details,
        ipAddress: ipAddress || null
      }
    });
  } catch (err) {
    console.error(`[AUDIT_LOG_ERROR] Failed to record audit log for action: ${action}`, err.message);
    return null;
  }
}

module.exports = {
  logAudit
};

File: backend/app/services/audit_service.py
import json
from typing import Optional, Any
from sqlalchemy.orm import Session
from app import models

def log_audit(
    db: Session,
    user_id: Optional[str] = None,
    action: str = "",
    resource: str = "",
    details: Optional[Any] = None,
    ip_address: Optional[str] = None
) -> Optional[models.AuditLog]:
    try:
        details_str = json.dumps(details) if isinstance(details, (dict, list)) else (str(details) if details is not None else None)
        log = models.AuditLog(
            userId=user_id,
            action=action,
            resource=resource,
            details=details_str,
            ipAddress=ip_address
        )
        db.add(log)
        db.flush()
        return log
    except Exception as err:
        print(f"[AUDIT_LOG_ERROR] Failed to record audit log for action: {action} - {err}")
        return None
